File: 1450-delete-leaves-with-a-given-value/delete-leaves-with-a-given-value.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def removeLeafNodes(self, root: Optional[TreeNode], target: int) -> Optional[TreeNode]:
        # POST ORDER TRAVERSAL REMOVAL METHOD 
        stack = [root]
        visit = set()
        parents = { root : None }

        while stack:
            node = stack.pop()

            if not node.right and not node.left and target == node.val:
                p = parents[node]

                if not p:
                    return None

                # PART EXACTLY WHERE THE NODE GETS UNLINKED (DELETED) 

                if p and p.left == node:
                    p.left = None 
                if p and p.right == node:
                    p.right = None

            elif node not in visit:
                visit.add(node)
                stack.append(node)

                if node.left:
                    stack.append(node.left)
                    parents[node.left] = node 
                if node.right:
                    stack.append(node.right)
                    parents[node.right] = node
        return root

                
                



        # RECURSIVE APPROACH 

        if not root:
            return None 
        
        root.left = self.removeLeafNodes(root.left , target)
        root.right = self.removeLeafNodes(root.right , target)


        if not root.left and not root.right and target == root.val:
            return None
        
        return root
        
        

File: 1450-delete-leaves-with-a-given-value/delete-leaves-with-a-given-value.java
class Solution
{
    public TreeNode removeLeafNodes(TreeNode root, int target)
    {
        if (root == null)
        {
            return null;
        }

        Deque<TreeNode> stack = new ArrayDeque<>();
        Set<TreeNode> visit = new HashSet<>();
        Map<TreeNode, TreeNode> parents = new HashMap<>();

        stack.push(root);
        parents.put(root, null);

        while (!stack.isEmpty())
        {
            TreeNode node = stack.peek();

            if (node.left == null && node.right == null && node.val == target)
            {
                stack.pop();
                TreeNode p = parents.get(node);

                if (p == null)
                {
                    return null;
                }

                if (p.left == node)
                {
                    p.left = null;
                }
                if (p.right == node)
                {
                    p.right = null;
                }
            }
            else if (!visit.contains(node))
            {
                visit.add(node);

                if (node.left != null)
                {
                    stack.push(node.left);
                    parents.put(node.left, node);
                }
                if (node.right != null)
                {
                    stack.push(node.right);
                    parents.put(node.right, node);
                }
            }
            else
            {
                stack.pop();
            }
        }
        return root;
    }
}